src/Workspaces/Core/MSBuild.BuildHost/MonoMSBuildDiscovery.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost;
+
+internal static class MonoMSBuildDiscovery
+{
+    private static IEnumerable<string>? s_searchPaths;
+    private static string? s_monoRuntimeExecutablePath;
+    private static string? s_monoLibDirPath;
+    private static string? s_monoMSBuildDirectory;
+
+    private static IEnumerable<string> GetSearchPaths()
+    {
+        if (s_searchPaths == null)
+        {
+            var path = Environment.GetEnvironmentVariable("PATH");
+            if (path == null)
+            {
+                return Array.Empty<string>();
+            }
+
+            s_searchPaths = path
+                .Split(Path.PathSeparator)
+                .Select(p => p.Trim('"'));
+        }
+
+        return s_searchPaths;
+    }
+
+    // http://man7.org/linux/man-pages/man3/realpath.3.html
+    // CharSet.Ansi is UTF8 on Unix
+    [DllImport("libc", EntryPoint = "realpath", CharSet = CharSet.Ansi, ExactSpelling = true, CallingConvention = CallingConvention.Cdecl)]
+    private static extern IntPtr Unix_realpath(string path, IntPtr buffer);
+
+    // http://man7.org/linux/man-pages/man3/free.3.html
+    [DllImport("libc", EntryPoint = "free", ExactSpelling = true, CallingConvention = CallingConvention.Cdecl)]
+    private static extern void Unix_free(IntPtr ptr);
+
+    /// <summary>
+    /// Returns the canonicalized absolute path from a given path, expanding symbolic links and resolving
+    /// references to /./, /../ and extra '/' path characters.
+    /// </summary>
+    private static string? RealPath(string path)
+    {
+        if (PlatformInformation.IsWindows)
+        {
+            throw new PlatformNotSupportedException($"{nameof(RealPath)} can only be called on Unix.");
+        }
+
+        var ptr = Unix_realpath(path, IntPtr.Zero);
+        var result = Marshal.PtrToStringAnsi(ptr); // uses UTF8 on Unix
+        Unix_free(ptr);
+
+        return result;
+    }
+
+    /// <summary>
+    /// Returns the fully qualified path to the mono executable.
+    /// </summary>
+    private static string? GetMonoRuntimeExecutablePath()
+    {
+        if (PlatformInformation.IsWindows)
+        {
+            return null;
```

Comment:
should these checks be throwing instead?  We should never get here on windows I assume

Reply:
Yeah, that's reasonable; if we end up here on Windows something went wrong. We probably could detect mono if it's installed, but until we write that code this is just bad.

Reply:
Changed!